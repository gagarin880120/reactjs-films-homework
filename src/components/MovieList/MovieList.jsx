import React from 'react';
import styles from './MovieList.module.scss';
import MovieCardContainer from '../MovieCard/MovieCardContainer';
import TrailerModalContainer from '../TrailerModal/TrailerModalContainer';

export default function MovieList(props) {
  return (
    <div className={styles.wrapper}>
      {props.results.map((result) => (
        <MovieCardContainer
          key={result.id}
          movie={result}
        />
      ))}
      {
        props.isModalOpen ? <TrailerModalContainer /> : null
      }

    </div>
  )
}
