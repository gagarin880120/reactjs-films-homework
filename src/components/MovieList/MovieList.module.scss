.wrapper {
  background: #f5f5f5;
  padding: 2%;
}

.container {
  margin-top: 2%;
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  column-gap: 10px;
  row-gap: 40px;
  justify-items: center;
}

@media screen and (min-width: 1600px) {
  .container {
    grid-template-columns: repeat(5, 1fr);
  }
}

@media screen and (max-width: 1200px) {
  .container {
    padding: 1%;
    grid-template-columns: repeat(3, 1fr);
  }
}

@media screen and (max-width: 890px) {
  .container {
    padding: 1%;
    grid-template-columns: repeat(2, 1fr);
  }
}

@media screen and (max-width: 600px) {
  .container {
    padding: 1%;
    grid-template-columns: 1fr;
  }
}
