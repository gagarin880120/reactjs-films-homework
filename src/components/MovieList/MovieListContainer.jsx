import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieList from './MovieList';
import { resultsSelector, modalSelector, filteredResultsSelector } from '../../redux/selectors';

export function MovieListContainer({ results, isModalOpen }) {
  return (
    <MovieList
      results={results}
      isModalOpen={isModalOpen}
    />
  );
}

export const mapStateToProps = (state) => ({
  results: filteredResultsSelector(state) ? filteredResultsSelector(state) : resultsSelector(state),
  isModalOpen: modalSelector(state),
});

MovieListContainer.propTypes = {
  results: PropTypes.instanceOf(Array),
  isModalOpen: PropTypes.bool,
};

MovieListContainer.defaultProps = {
  results: [],
  isModalOpen: false,
};

export default connect(mapStateToProps)(MovieListContainer);
