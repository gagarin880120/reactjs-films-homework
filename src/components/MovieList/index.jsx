import React, { useEffect, useMemo, createRef, useState } from 'react';
import MovieList from './MovieList';
import InfoButton from '../InfoButton/InfoButton';
import TrailerButtonRect from '../TrailerButtonRect/TrailerButtonRect';
import TrailerButtonRound from '../TrailerButtonRound/TrailerButtonRound';
import styles from './MovieList.module.scss';
import { useSelector, useDispatch } from 'react-redux';
import { searchResultsSelector, genresSelector } from '../../redux/selectors';
import { getGenres } from '../../redux/actions';
import { getGenresString, getTrailer } from '../../helpers/helpers';

export default function MovieListContainer() {
  const [trailerSource, setTrailerSource] = useState('');
  const [isTrailerLoading, setIsTrailerLoading] = useState(false);
  const [modalIsOpen, setIsOpen] = useState(false);
  const [isInfoViewed, setIsInfoViewed] = useState(Array.from({length: 20}, () => false));
  const hoverRefs= useMemo(
    () => Array.from({ length: 20 }, () => createRef()),
    []
  );
  const infoRefs= useMemo(
    () => Array.from({ length: 20 }, () => createRef()),
    []
  );
  const results = useSelector(searchResultsSelector);
  const genres = useSelector(genresSelector);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getGenres());
  }, [])

  const arrayOfResults = results.map((result, i) => {
    return (
      <div className={styles.movieCard}
        key={i}
        onMouseOver={(e) => {
          hoverRefs[i].current.style.display = isInfoViewed[i] ? 'none' : 'flex';
        }}
        onMouseLeave={(e) => {
          hoverRefs[i].current.style.display = 'none';
        }}
        style={{
          backgroundImage: result.poster_path ?
          `url(https://image.tmdb.org/t/p/w300/${result.poster_path})`:
          'url(https://spnbag.com/assets/general/images/no_poster.jpg)'
        }}
      >

        <div
          ref={hoverRefs[i]}
          className={styles.hoverTrailer}
        >
          <TrailerButtonRound onTrailerButtonClick={() =>
            {
              setIsTrailerLoading(true);
              getTrailer(result.id).then(data => {
                setTrailerSource(data.videos.results[0].key)
              })
              .catch(() => setTrailerSource(null))
              .finally(() => {
                setIsTrailerLoading(false);
              });
              setIsOpen(true);
            }} />
          <span className={styles.hoverTrailerText}>Watch Now</span>
          <InfoButton onInfoButtonClick={() => {
            hoverRefs[i].current.style.display = 'none';
            setIsInfoViewed(prevState => prevState.map((v, idx) => idx === i ? true : v));
            infoRefs[i].current.className = styles.infoIsViewed;
          }} />
        </div>

        <div className={styles.movieInfo} ref={infoRefs[i]}>
          {
            isInfoViewed[i] ? <div className={styles.closeInfo} onClick={() => {
              setIsInfoViewed(prevState => prevState.map((v, idx) => idx === i ? false : v));
              infoRefs[i].current.className = styles.movieInfo;
            }}><span>X</span></div> : null
          }
          <span className={styles.movieTitle}>
            {
            result.title.length < 18 ?
              result.title :
              result.title.slice(0, 18) + '...'
            }
          </span>
          <span className={styles.movieRating}>{result.vote_average}</span>
          <p className={styles.movieGenre}>{getGenresString(result.genre_ids, genres)}</p>
          {
            isInfoViewed[i] ?
            <>
              <div className={styles.overview}>{result.overview}</div>
              <TrailerButtonRect onTrailerButtonClick={() =>
            {
              getTrailer(result.id, setTrailerSource, setIsTrailerLoading);
              setIsOpen(true);
            }}/>
            </> : null
          }
        </div>
      </div>
    )
  })

  return (
    <div>
      <MovieList
      arrayOfResults={arrayOfResults}
      />
      <div
        className={styles.modal}
        style={{display: modalIsOpen ? 'flex' : 'none'}}
        onClick={() => setIsOpen(false)}
      >
        {isTrailerLoading ? null : trailerSource ?
          <iframe
          width="800"
          height="450"
          src={`https://www.youtube.com/embed/${trailerSource}`}
          frameBorder="0"
          allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen>
          </iframe> :
          <span className={styles.noTrailer}>
            Trailer is not available
          </span>
        }

      </div>
    </div>
  )
}
