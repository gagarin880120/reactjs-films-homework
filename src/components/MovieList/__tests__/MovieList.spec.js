import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import MovieList from '../MovieList';

const results= [
  {"popularity":48.192,
  "vote_count":16954,
  "video":false,
  "poster_path":"/f89U3ADr1oiB1s9GkdPOEpXUk5H.jpg",
  "id":603,
  "adult":false,
  "backdrop_path":"/ByDf0zjLSumz1MP1cDEo2JWVtU.jpg",
  "original_language":"en","original_title":"The Matrix",
  "genre_ids":[28,878],
  "title":"The Matrix",
  "vote_average":8.1,
  "overview":"Set in the 22nd century, The Matrix tells the story of a computer hacker...",
  "release_date":"1999-03-30"},
  {"popularity":26.959,
  "vote_count":6820,
  "video":false,
  "poster_path":"/aA5qHS0FbSXO8PxcxUIHbDrJyuh.jpg",
  "id":604,
  "adult":false,
  "backdrop_path":"/kQIzFGdVCwsuutMKyCVZ8LrJw3e.jpg",
  "original_language":"en",
  "original_title":"The Matrix Reloaded",
  "genre_ids":[28,12,878,53],
  "title":"The Matrix Reloaded",
  "vote_average":6.9,
  "overview":"Six months after the events depicted in The Matrix, Neo has proved...",
  "release_date":"2003-05-15"}
]

test('MovieList component should render without crashing', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<MovieList results={results} />);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
