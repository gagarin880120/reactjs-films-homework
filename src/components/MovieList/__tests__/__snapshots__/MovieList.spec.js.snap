// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component should render without crashing 1`] = `
<div
  className="wrapper"
>
  <MovieCard
    genre_ids={
      Array [
        28,
        878,
      ]
    }
    id={603}
    overview="Set in the 22nd century, The Matrix tells the story of a computer hacker..."
    poster_path="/f89U3ADr1oiB1s9GkdPOEpXUk5H.jpg"
    title="The Matrix"
    vote_average={8.1}
  />
  <MovieCard
    genre_ids={
      Array [
        28,
        12,
        878,
        53,
      ]
    }
    id={604}
    overview="Six months after the events depicted in The Matrix, Neo has proved..."
    poster_path="/aA5qHS0FbSXO8PxcxUIHbDrJyuh.jpg"
    title="The Matrix Reloaded"
    vote_average={6.9}
  />
</div>
`;
